ls1intum/Artemis
# [Feature] Handle Illegal Exam Programming Submissions

### Checklist

### Motivation and Context


Follow-up to #3060

In case a student has still the write permission of his repository of the exam programming exercise, a student is able to submit code and receive a valid and rated result. 

### Description

In this follow-up we address, we address the changes to SQL queries and thereof the assessment of illegal submissions. 
In case a tutor comes across an illegal submission during the assessment of exam programming exercises, a warning is shown.
In all other cases where we fetch the Submissions of a Participation or the Result of a Participation, we filter out all connected to illegal Submissions. 

The data structure of affected entities is the following (bold entities indicate possible illegal roots):
Participation -> **Submissions** -> **Result**
Participation -> **Results**

All results which are created and attached to the latest Submission, are also attached to the overall results of a Participation.

We adapt a lot of SQL queries and modify them by adding a filter for Illegal Submissions. We also change the function names by appending the word 'legal'.

### Steps for Testing

This feature is currently only supported for programming exercises in Exams. 

1. Log in to Artemis
2. Navigate to Course Administration 
3. Create an exam with one or more programming exercises
4. Register one or more students and have them participate in the exam by submitting code
5. Wait until the exam ends
6. Navigate to the submission page and take a look at the submissions. There should be no illegal ones.
7. Now navigate to student's exercise code repository on either Bitbucket or Gitlab 
8. Edit a code file. This will generate a new illegal submission in Artemis
9. Wait a bit and go to the submission page and make sure that the last submission is of type "ILLEGAL"

Assessing the exercise as a tutor or instructor
1. Make sure you set the programming exercise to manual review
2. Navigate to Course Administration -> Exams -> Exercise Groups -> Submissions -> Assess submission 
3. Make sure that the red warning is displayed on the top left
4. Make sure the code editor and test scores correspond to that submission

### Test Coverage



ProgrammingSubmissionsService.java 75%


### Screenshots


The submission page shows whether a submission is legal or not:

![image](https://user-images.githubusercontent.com/11298674/113307720-beca9880-9305-11eb-8360-de65ff843c22.png)

Instructors/Tutors will get a warning if they assess an illegal submission:

![image](https://user-images.githubusercontent.com/11298674/113307889-e7eb2900-9305-11eb-9bd7-00431f3071eb.png)
